### PR TITLE
Implemented yielding Dir.each_child

### PR DESCRIPTION
**Pull request by Sija**
Complements #4808

### CODE REVIEW

**Comment by Sija** [comment]
Ooops, build on 32-bit linux box failed with:

```
Invalid memory access (signal 11) at address 0x25eeb4d0

[0x902449a] *CallStack::print_backtrace:Int32 +154
[0x893900f] __crystal_sigfault_handler +111
[0xbc4f3f1] sigfault_handler +37
[0xf77b0cc0] ???
[0xbb5a572] _ZN4llvm23ReplaceableMetadataImpl7dropRefEPv +50
[0xbb5a65c] _ZN4llvm16MetadataTracking7untrackEPvRNS_8MetadataE +44
[0xbb5dd90] _ZN4llvm6MDNode10setOperandEjPNS_8MetadataE +64
[0xbb5dfcf] _ZN4llvm6MDNode17dropAllReferencesEv +47
[0xbb4644d] _ZN4llvm15LLVMContextImplD1Ev +11389
[0xbb3e8c3] _ZN4llvm11LLVMContextD1Ev +35
[0xba59fd1] LLVMContextDispose +33
[0xa55a398] *LLVM::Context#finalize:Nil +184
[0x8946b9e] ~proc7Proc(Pointer(Void), Pointer(Void), Nil) +30
[0xbc3edce] GC_invoke_finalizers +110
[0xbc3f036] GC_notify_or_invoke_finalizers +198
[0xbc40a51] GC_generic_malloc_many +81
[0xbc492ed] GC_malloc +205
[0x891b5ce] __crystal_malloc +30
[0x9e814f5] *Crystal::NumberLiteral::new<String, Symbol>:Crystal::NumberLiteral +53
[0xa2f4297] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +1159
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa30706c] *Crystal::Parser#parse_op_assign_no_control<Bool, Bool>:Crystal::ASTNode+ +92
[0xa306ffc] *Crystal::Parser#parse_op_assign_no_control:Crystal::ASTNode+ +76
[0xa330f0d] *Crystal::Parser#parse_call_arg:Crystal::ASTNode+ +317
[0xa317b3c] *Crystal::Parser#parse_call_args<Bool, Bool, Bool>:(Crystal::Parser::CallArgs | Nil) +1724
[0xa3173ca] *Crystal::Parser#parse_call_args:stop_on_do_after_space<Bool>:(Crystal::Parser::CallArgs | Nil) +122
[0xa315035] *Crystal::Parser#parse_var_or_call<Bool, Bool>:Crystal::ASTNode+ +1701
[0xa344ddc] *Crystal::Parser#parse_var_or_call:Crystal::ASTNode+ +76
[0xa2facc4] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +28340
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa2f01ac] *Crystal::Parser#parse_op_assign:Crystal::ASTNode+ +76
[0xa2f0133] *Crystal::Parser#parse_expression:Crystal::ASTNode+ +67
[0xa2ef6d0] *Crystal::Parser#parse_multi_assign:Crystal::ASTNode+ +96
[0xa2ef008] *Crystal::Parser#parse_expressions_internal:Crystal::ASTNode+ +248
[0xa2eeeeb] *Crystal::Parser#parse_expressions:Crystal::ASTNode+ +75
[0xa333e96] *Crystal::Parser#parse_begin:Crystal::ASTNode+ +86
[0xa2f5283] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +5235
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa30706c] *Crystal::Parser#parse_op_assign_no_control<Bool, Bool>:Crystal::ASTNode+ +92
[0xa306ffc] *Crystal::Parser#parse_op_assign_no_control:Crystal::ASTNode+ +76
[0xa2f15c1] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +5121
[0xa2f01ac] *Crystal::Parser#parse_op_assign:Crystal::ASTNode+ +76
[0xa2f0133] *Crystal::Parser#parse_expression:Crystal::ASTNode+ +67
[0xa2ef6d0] *Crystal::Parser#parse_multi_assign:Crystal::ASTNode+ +96
[0xa2eef60] *Crystal::Parser#parse_expressions_internal:Crystal::ASTNode+ +80
[0xa2eeeeb] *Crystal::Parser#parse_expressions:Crystal::ASTNode+ +75
[0xa33731e] *Crystal::Parser#parse_def_helper<Bool, Bool>:Crystal::Def +6462
[0xa33aa46] *Crystal::Parser#parse_def<Bool, Bool, Nil>:Crystal::Def +134
[0xa33a9ac] *Crystal::Parser#parse_def:Crystal::Def +76
[0xa2f688d] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +10877
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa2f01ac] *Crystal::Parser#parse_op_assign:Crystal::ASTNode+ +76
[0xa34406e] *Crystal::Parser#parse_visibility_modifier<Crystal::Visibility>:Crystal::VisibilityModifier +190
[0xa2fa881] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +27249
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa2f01ac] *Crystal::Parser#parse_op_assign:Crystal::ASTNode+ +76
[0xa2f0133] *Crystal::Parser#parse_expression:Crystal::ASTNode+ +67
[0xa2ef6d0] *Crystal::Parser#parse_multi_assign:Crystal::ASTNode+ +96
[0xa2ef008] *Crystal::Parser#parse_expressions_internal:Crystal::ASTNode+ +248
[0xa2eeeeb] *Crystal::Parser#parse_expressions:Crystal::ASTNode+ +75
[0xa34007c] *Crystal::Parser#parse_module_def:Crystal::ModuleDef +300
[0xa2f85fd] *Crystal::Parser#parse_atomic_without_location:Crystal::ASTNode+ +18413
[0xa2f3d73] *Crystal::Parser#parse_atomic:Crystal::ASTNode+ +67
[0xa2f3d03] *Crystal::Parser#parse_atomic_with_method:Crystal::ASTNode+ +67
[0xa2f3837] *Crystal::Parser#parse_prefix:Crystal::ASTNode+ +215
[0xa2f3556] *Crystal::Parser#parse_pow:Crystal::ASTNode+ +70
[0xa2f325c] *Crystal::Parser#parse_mul_or_div:Crystal::ASTNode+ +76
[0xa2f2bdf] *Crystal::Parser#parse_add_or_sub:Crystal::ASTNode+ +95
[0xa2f2936] *Crystal::Parser#parse_shift:Crystal::ASTNode+ +70
[0xa2f26e6] *Crystal::Parser#parse_logical_and:Crystal::ASTNode+ +70
[0xa2f2456] *Crystal::Parser#parse_logical_or:Crystal::ASTNode+ +70
[0xa2f207c] *Crystal::Parser#parse_cmp:Crystal::ASTNode+ +76
[0xa2f1c9c] *Crystal::Parser#parse_equality:Crystal::ASTNode+ +76
[0xa2f1af3] *Crystal::Parser#parse_and:Crystal::ASTNode+ +67
[0xa2f1953] *Crystal::Parser#parse_or:Crystal::ASTNode+ +67
[0xa2f1824] *Crystal::Parser#parse_range:Crystal::ASTNode+ +68
[0xa2f169c] *Crystal::Parser#parse_question_colon:Crystal::ASTNode+ +44
[0xa2f0281] *Crystal::Parser#parse_op_assign<Bool, Bool>:Crystal::ASTNode+ +193
[0xa2f01ac] *Crystal::Parser#parse_op_assign:Crystal::ASTNode+ +76
[0xa2f0133] *Crystal::Parser#parse_expression:Crystal::ASTNode+ +67
[0xa2ef6d0] *Crystal::Parser#parse_multi_assign:Crystal::ASTNode+ +96
[0xa2eef60] *Crystal::Parser#parse_expressions_internal:Crystal::ASTNode+ +80
[0xa2eeeeb] *Crystal::Parser#parse_expressions:Crystal::ASTNode+ +75
[0xa2da4fc] *Crystal::Parser#parse:Crystal::ASTNode+ +60
[0xa49d2d0] *Crystal::TopLevelVisitor +944
[0x97e2dd8] *Crystal::ASTNode+ +2200
[0xa498cc9] *Crystal::TopLevelVisitor#visit<Crystal::Expressions>:Bool +153
[0x97e29c4] *Crystal::ASTNode+ +1156
[0xa49d34b] *Crystal::TopLevelVisitor +1067
[0x97e2dd8] *Crystal::ASTNode+ +2200
[0xa498cc9] *Crystal::TopLevelVisitor#visit<Crystal::Expressions>:Bool +153
[0x97e29c4] *Crystal::ASTNode+ +1156
[0xa49d34b] *Crystal::TopLevelVisitor +1067
[0x97e2dd8] *Crystal::ASTNode+ +2200
[0xa498cc9] *Crystal::TopLevelVisitor#visit<Crystal::Expressions>:Bool +153
[0x97e29c4] *Crystal::ASTNode+ +1156
[0x9290f3b] *Crystal::Program#top_level_semantic<Crystal::ASTNode+>:Tuple(Crystal::ASTNode+, Crystal::TypeDeclarationProcessor) +267
[0x9290591] *Crystal::Program#semantic<Crystal::ASTNode+, Bool>:Crystal::ASTNode+ +97
[0xa45c195] *Crystal::Compiler#compile<Crystal::Compiler::Source, String>:Crystal::Compiler::Result +405
[0x8966fef] *run<String, Nil, Bool, Crystal::Debug>:(Crystal::SpecRunOutput | LLVM::GenericValue) +735
[0x8966d02] *run<String>:(Crystal::SpecRunOutput | LLVM::GenericValue) +82
[0x89903cb] ~procProc(Nil) +43
[0x8965ed8] *it<String, String, Int32, Int32, &Proc(Nil)>:(Array(Spec::Result) | Nil) +600
[0x898ef76] ~procProc(Nil) +1142
[0x9237344] *Spec::RootContext::describe<String, String, Int32, &Proc(Nil)>:Spec::Context+ +484
[0x89667df] *describe<String, String, Int32, &Proc(Nil)>:Spec::Context+ +63
[0x891314d] ???
[0x8938d0d] main +77
[0xf6e8e637] __libc_start_main +247
[0x890ffab] ???
```

**Comment by ysbaddaden** [comment]
Are we hitting the memory limit of Travis 32-bit VM? Maybe we should build `std/compiler_spec.cr` and `std/std_spec.cr` one after the other.

**Comment by RX14** [comment]
@ysbaddaden I don't think that specific segfault is because of memory, but I may be proven wrong. Typically memory issues appear when `fork()` fails.

**Comment by straight-shoota** [comment]
Maybe initialize a few rebuilds to determine if the error is persistent?

**Comment by RX14** [comment]
@straight-shoota It's not, this error has been around for months.

**Comment by Sija** [comment]
@ysbaddaden @RX14 Coming back to the topic of this PR, shall we get this merged or shall I open another one, changing name conventions as noted in https://github.com/crystal-lang/crystal/pull/4808#issuecomment-321009687?

**Comment by RX14** [comment]
Well we need `each_child` anwyay, why not combine all these name changes into 1 PR? You should look at the instance methods too, see if any of those should be renamed.

**Comment by Sija** [comment]
@RX14 done.

**Comment by bew** [comment]
> You should look at the instance methods too, see if any of those should be renamed.

@Sija I think you missed that. Especially `Dir#each` should be renamed to `Dir#each_entry`.

Note: do we want to have `#each_child` `#entries` & `#children` instance methods as well?

**Comment by Sija** [comment]
> Note: do we want to have #each_child #entries & #children instance methods as well?

I'd say yes. WDYT @ysbaddaden @RX14 ?

**Comment by ysbaddaden** [comment]
Usually we'll want to return an array for easyness (e.g. to pass it along to a template), sometimes we'll want to yield a folder and don't need an array (avoids HEAP allocation), and we may need the full directory listing, including `.` and `..` once in a while (either returning an array or iterating).

**Comment by Sija** [comment]
@ysbaddaden said & done, ready for review :)

File: src/dir.cr [hunk 1]
```diff
@@ -305,6 +309,27 @@ class Dir
...
+    include Iterator(String)
+
+    def initialize(@dir : Dir)
+    end
+
+    def next
+      excluded = {".", ".."}
+      entry = nil
+      loop do
+        break unless excluded.includes?(entry = @dir.read)
```
Comment (RX14) [inline src/dir.cr]: Why not use early return here?
```cr
excluded = {".", ".."}
loop do
  entry = @dir.read
  return entry unless excluded.includes? entry
end
```

Comment (Sija) [inline src/dir.cr]: fixed.

Comment (bew) [inline src/dir.cr]: Can I suggest another one?
```cr
excluded = {".", ".."}
while entry = @dir.read
  return entry unless excluded.includes? entry
end
stop
```
Nevermind you don't have to include that one, just wanted to write it somewhere..

Edit: oh in fact you implemented it that way too, perfect then 😆

**Comment by Sija** [comment]
Anything more to do here, or is it GTG?

**Comment by RX14** [comment]
@Sija, @ysbaddaden hasn't reviewed the latest changes.

**Comment by Sija** [comment]
@ysbaddaden ping

**Comment by ysbaddaden** [comment]
Thank you!

**Comment by Sija** [comment]
still awesome experience, so yw, thx!

**Pull request merged**